certificate-authority/service/grpc/server.go

```diff
@@ -1,60 +1,112 @@
 package grpc
 
 import (
-	"context"
-	"crypto"
-	"crypto/x509"
-	"time"
+	"bytes"
+	"fmt"
 
-	"github.com/karrick/tparse/v2"
 	"github.com/plgd-dev/hub/v2/certificate-authority/pb"
 	"github.com/plgd-dev/hub/v2/certificate-authority/store/mongodb"
+	"github.com/plgd-dev/hub/v2/pkg/fn"
+	"github.com/plgd-dev/hub/v2/pkg/fsnotify"
 	"github.com/plgd-dev/hub/v2/pkg/log"
-	"github.com/plgd-dev/kit/v2/security"
+	"go.uber.org/atomic"
 )
 
-type CertificateSigner interface {
-	// csr is encoded by PEM and returns PEM
-	Sign(ctx context.Context, csr []byte) ([]byte, error)
-}
-
 // CertificateAuthorityServer handles incoming requests.
 type CertificateAuthorityServer struct {
 	pb.UnimplementedCertificateAuthorityServer
 
-	validFrom    func() time.Time
-	validFor     time.Duration
-	certificate  []*x509.Certificate
-	privateKey   crypto.PrivateKey
-	signerConfig SignerConfig
-	logger       log.Logger
-	ownerClaim   string
-	store        *mongodb.Store
-	hubID        string
+	signerConfig     SignerConfig
+	logger           log.Logger
+	ownerClaim       string
+	store            *mongodb.Store
+	hubID            string
+	fileWatcher      *fsnotify.Watcher
+	onFileChangeFunc func(event fsnotify.Event)
+
+	signer atomic.Pointer[Signer]
 }
 
-func NewCertificateAuthorityServer(ownerClaim string, hubID string, signerConfig SignerConfig, store *mongodb.Store, logger log.Logger) (*CertificateAuthorityServer, error) {
-	certificate, err := security.LoadX509(signerConfig.CertFile)
+func NewCertificateAuthorityServer(ownerClaim string, hubID string, signerConfig SignerConfig, store *mongodb.Store, fileWatcher *fsnotify.Watcher, logger log.Logger) (*CertificateAuthorityServer, error) {
+	s := &CertificateAuthorityServer{
+		signerConfig: signerConfig,
+		logger:       logger,
+		ownerClaim:   ownerClaim,
+		store:        store,
+		hubID:        hubID,
+		fileWatcher:  fileWatcher,
+	}
+
+	_, err := s.load()
 	if err != nil {
 		return nil, err
 	}
-	privateKey, err := security.LoadX509PrivateKey(signerConfig.KeyFile)
+
+	var removeFilesOnError fn.FuncList
+	for _, ca := range signerConfig.caPoolArray {
+		if err := fileWatcher.Add(ca); err != nil {
+			removeFilesOnError.Execute()
+			return nil, fmt.Errorf("cannot watch CAPool(%v): %w", ca, err)
+		}
+		removeFilesOnError.AddFunc(func() {
+			_ = fileWatcher.Remove(ca)
```

Comment:
if I recall corectly Go captures variables as references in closures, so all functions added here will have the same `ca`.

Assigning to helper variable should work here:

```Go
for _, ca := range signerConfig.caPoolArray {
  ...
  caToRemove := ca
  removeFilesOnError.AddFunc(func() {
    _ = fileWatcher.Remove(caToRemove)
  }
}

```

Reply:
Good point